content/capabilities/single-use.md

```diff
@@ -0,0 +1,139 @@
+---
+title: "Single Use Object Capabilities"
+section: "Capabilities"
+menu:
+  toc:
+    parent: "capabilities"
+    weight: 10
+---
+## Problem
+
+As shown in the [tutorial page about object capabilities](https://tutorial.ponylang.io/object-capabilities/object-capabilities.html), we can limit the actions of other objects or actors with object capabilities. This is used in the standard library of Pony for network connections and file access, for example. But it can also be used for other systems created in Pony.
+
+For example, let's say we want to implement a capability-restricted service, that returns one unique number every time it is called. We will use a `CustomAuth` as a token to restrict its access, created when provided `AmbientAuth`:
+
+```pony
+use "promises"
+
+primitive CustomAuth
+  new create(auth: AmbientAuth) => None
+
+actor RestrictedService
+  var current_count: USize = 1
+
+  be apply(auth: CustomAuth, promise: Promise[USize]) =>
+    promise(current_count = current_count + 1)
+```
+
+Our `Main` actor receives the `AmbientAuth` token on creation from `env.root`, which means only it or something it provided with that capability can create a new `CustomAuth`. We can then hand out these `CustomAuth` tokens to other actors or objects that need to call our `RestrictedService`.
+
+However, let's suppose that we want these other actors to only call this restricted service _once_. As it currently stands, nothing prevents them from calling `RestrictedService.apply` several times, thus requesting a new number. The current object capability example doesn't allow us to limit how many times the token can be used.
+
+You might think that tracking every caller of the service with a `HashMap` of identities could solve this problem, but keep in mind that not only is this cumbersome, but it can also be circumvented if using, say, interfaces as a parameter to the call.
```

Comment:
It may also be worth pointing out there is a runtime cost to these kinds of hash map lookups, which is likely greater than the runtime cost of the pattern proposed in this document.

There's also nothing stopping the caller from lying about their identity, since the only way to get the identity of a caller is to have the caller pass a reference to themselves as an argument. They could easily construct and use a new "throwaway" identity object to pass in as the supposed identity of the caller. This circumvention could be mitigated by using a direct actor callback for the response path instead of a promise - the identity can't be faked if you need to use it as the "return address" of the message.

On that note, I don't really understand the circumvention risk you mentioned here:
> but it can also be circumvented if using, say, interfaces as a parameter to the call

Reply:
I guess I wasn't very clear regarding this restriction. The point would be to limit the call to only one actor/object by passing `this` to the service, and allow any actor to do that by having, say, a limited or empty interface/trait that the actor must implement. However, an actor can just create another actor/object that implements the interface/trait to pass as that parameter, thus making multiple calls.

Perhaps I can simplify this simple example to only consider a specific type for the caller, or just clarify this point further.